contracts/EngineRegistryV0.sol

```diff
@@ -0,0 +1,71 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.17;
+
+import "./interfaces/0.8.x/IEngineRegistryV0.sol";
+import "@openzeppelin-4.7/contracts/utils/introspection/ERC165.sol";
+
+/**
+ * @title Engine Registry contract, V0.
+ * @author Art Blocks Inc.
+ * @notice Privileged Roles and Ownership:
+ * This contract intentionally has no owners or admins, and is intended
+ * to be deployed with a permissioned `deployerAddress` that may speak
+ * to this registry. If in the future multiple deployer addresses are
+ * needed to interact with this registry, a new registry version with
+ * more complex logic should be implemented and deployed to replace this.
+ */
+contract EngineRegistryV0 is IEngineRegistryV0, ERC165 {
+    /// configuration variable (determined at time of deployment)
+    /// that determines what address may perform registration actions.
+    address internal immutable deployerAddress;
+
+    /// internal mapping for managing known list of registered contracts.
+    mapping(address => bool) internal registeredContractAddresses;
+
+    constructor(address _deployerAddress) {
+        deployerAddress = _deployerAddress;
+    }
+
+    /**
+     * @inheritdoc IEngineRegistryV0
+     */
+    function registerContract(
+        address _contractAddress,
+        string calldata _coreVersion,
+        string calldata _coreType
+    ) external {
+        // CHECKS
+        // Validate against `tx.origin` rather than `msg.sender` as it is intended that this registration be
+        // performed in an automated fashion *at the time* of contract deployment for the `_contractAddress`.
+        require(
+            tx.origin == deployerAddress,
+            "Only allowed deployer-address TX origin"
+        );
+
+        // EFFECTS
+        emit ContractRegistered(_contractAddress, _coreVersion, _coreType);
+        registeredContractAddresses[_contractAddress] = true;
+    }
+
+    /**
+     * @inheritdoc IEngineRegistryV0
+     */
+    function unregisterContract(address _contractAddress) external {
+        // CHECKS
+        // Validate against `tx.origin` rather than `msg.sender` for consistency with the above approach,
+        // as we expect in usage of this contract `msg.sender == tx.origin` to be a true assessment.
```

Comment:
super super minor nit, but I feel like the `msg.sender == tx.origin` is an important assumption in this context and  makes sense to highlight just in the above check comments (in registerContract), since you're referencing them here anyway

Reply:
Ah – well in the case of the above callsite, this actually isn't true that we expect `msg.sender == tx.origin`, we actually only expect it typically in the `unregisterContract` path. In the `registerContract` path, we expect `msg.sender != tx.origin`, and to intentionally need to validate against `tx.origin.

Reply:
ah got it, appreciate the clarification here.

more so for my own understanding, is this because it's more likely that a registration call is potentially coming from a contract vs EOA, while we expect that not to be the case for unregisterContract calls (and expect those to primarily come from an EOA)? if so, why is that?

Reply:
sorry bit of a brain short circuit from my end... registration calls are pretty much solely coming from the constructor of the contracts being deployed by that deployer address, whereas unregistering would be a manual process so this all makes sense 👍

Reply:
Yep yep–not a worry ser; your EDIT response is spot on: registrations are expected to happen via contract-being-deployed auto-registering in the constructor, vs. unregistering likely being something that is manual (though we don't have current plans/expected cases of unregistrations), but even if it were automated it would need to be from a `tx.origin` of the allowed deployer.